Higher Order Components (HOC)

kLabz opened the pull request:
## Higher Order Components (HOC)

This approach aims to allow the creation of container components as described on [ReactRedux documentation](http://redux.js.org/docs/basics/UsageWithReact.html#implementing-container-components), as an alternative to the `IConnectedComponent` approach.

The HOC approach (already mentioned as a TODO in the current readme) seem closer to the usual React approach, and may make transitioning from React/Redux to haxe-react/haxe-redux easier.

The separation between the HOC and its wrapped component is made clear here, since the HOC can only do its container role, and nothing else. This method may have some positive impact on performance too, since it will only add what's needed through `ReactRedux.connect()` (mapState and/or mapDispatch).

Many compile-time checks are done to ensure typing is correct. Compilation warnings / errors will be issued depending on the issue.

### VisibleTodoList example

#### Haxe version

```haxe
// A ReactConnector providing `TodoListProps`, with `TodoList` as default wrapped component
class VisibleTodoList extends ReactConnector<TodoList, TodoListProps>
{
	static function mapStateToProps(state:ApplicationState, ownProps:Dynamic):Partial<TodoListProps>
	{
		return {
			todos: getVisibleTodos(state.todoList.todos, state.todoList.visibilityFilter)
		}
	}

	static function getVisibleTodos(todos:Array<TodoData>, filter:TodoFilter)
	{
		return switch (filter) {
			case SHOW_ALL: todos;
			case SHOW_COMPLETED: todos.filter(function(t) return t.completed);
			case SHOW_ACTIVE: todos.filter(function(t) return !t.completed);
		}
	}

	static function mapDispatchToProps(dispatch:Dispatch, ownProps:Dynamic):Partial<TodoListProps>
	{
		return {
			onTodoClick: function(id) {
				return dispatch(TodoAction.Toggle(id));
			}
		};
	}
}
```

```haxe
@:jsxStatic('render')
class TodoList
{
	static public function render(props:TodoListProps)
	{
		var todos = props.todos.map(function(todo) return jsx('
			<$Todo
				key=${todo.id}
				{...todo}
				onClick=${props.onTodoClick.bind(todo.id)}
			/>
		'));

		return jsx('
			<ul>
				$todos
			</ul>
		');
	}
}
```

You can then call the HOC like this:

```haxe
	override function render()
	{
		return jsx('<$VisibleTodoList />');
	}
```

#### Javascript version

http://redux.js.org/docs/basics/UsageWithReact.html#containersvisibletodolistjs

```javascript
const getVisibleTodos = (todos, filter) => {
	switch (filter) {
		case 'SHOW_ALL':
			return todos
		case 'SHOW_COMPLETED':
			return todos.filter(t => t.completed)
		case 'SHOW_ACTIVE':
			return todos.filter(t => !t.completed)
	}
}

const mapStateToProps = (state) => {
	return {
		todos: getVisibleTodos(state.todos, state.visibilityFilter)
	}
}

const mapDispatchToProps = (dispatch) => {
	return {
		onTodoClick: (id) => {
			dispatch(toggleTodo(id))
		}
	}
}

const VisibleTodoList = connect(
	mapStateToProps,
	mapDispatchToProps
)(TodoList)
```

http://redux.js.org/docs/basics/UsageWithReact.html#componentstodolistjs

```javascript
const TodoList = ({ todos, onTodoClick }) => (
	<ul>
		{todos.map(todo =>
			<Todo
				key={todo.id}
				{...todo}
				onClick={() => onTodoClick(todo.id)}
			/>
		)}
	</ul>
)

TodoList.propTypes = {
	todos: PropTypes.arrayOf(PropTypes.shape({
		id: PropTypes.number.isRequired,
		completed: PropTypes.bool.isRequired,
		text: PropTypes.string.isRequired
	}).isRequired).isRequired,
	onTodoClick: PropTypes.func.isRequired
}
```

### HOC's own props

If your HOC needs props, you can extend `ReactConnectorOfProps` instead of `ReactConnector`:

```haxe
class FilterLink extends ReactConnectorOfProps<Link, LinkProps, FilterLinkProps>
{
	static function mapStateToProps(state:ApplicationState, ownProps:FilterLinkProps):Partial<LinkProps>
	{
		return {
			active: ownProps.filter == state.todoList.visibilityFilter
		}
	}

	static function mapDispatchToProps(dispatch:Dispatch, ownProps:FilterLinkProps):Partial<LinkProps>
	{
		return {
			onClick: function() {
				return dispatch(TodoAction.SetVisibilityFilter(ownProps.filter));
			}
		};
	}
}
```

You can then use it like this:

```haxe
jsx('
	<ul>
		<li>
			<$FilterLink filter="SHOW_ALL" label="All" />
		</li>
		<li>
			<$FilterLink filter="SHOW_ACTIVE" label="Active" />
		</li>
		<li>
			<$FilterLink filter="SHOW_COMPLETED" label="Completed" />
		</li>
	</ul>
');
```

### Reusing the HOC with another wrapped component

In the above example, you can also use another component instead of `TodoList`, by using it as `children`:

```haxe
	override function render()
	{
		return jsx('
			<$VisibleTodoList>
				<$MyOtherTodoList />
			</$VisibleTodoList>
		');
	}
```

The generated props would be added to every direct child's props, and will override the default wrapped component.

You can also create HOCs without binding them to specific components, with `ReactGenericConnector<TProps>` or `ReactGenericConnectorOfProps<TProps, TOwnProps>`.
They will return an empty node at runtime if you do not provide any children, and issue a `console.error()` if you compiled with `-debug`.

### HOC composition

#### Composition hardcoded inside a connector

You can use a connector accepting ownProps as `TComponent` for another connector.

#### Composition via jsx

```haxe
jsx('
	<$FirstConnector>
		<$SecondConnector>
			<$Component />
		</$SecondConnector>
	</$FirstConnector>
');
```

### ReactConnector API

```haxe
/**
  A react-redux connector designed for a component TComponent, providing TProps
  from redux store.
*/
typedef ReactConnector<TComponent, TProps> = ReactConnect<TComponent, TProps, TVoid>;

/**
  A react-redux connector designed for a component TComponent, providing TProps
  from redux store and the connector's props (TOwnProps)
*/
typedef ReactConnectorOfProps<TComponent, TComponentProps, TOwnProps> = ReactConnect<TComponent, TComponentProps, TOwnProps>;

/**
  A react-redux connector for use with any component compatible with TProps.
  Will return an empty react node at runtime if used without children, with an
  error message if compiled with -debug.
*/
typedef ReactGenericConnector<TProps> = ReactConnect<TVoid, TProps, TVoid>;

/**
  A react-redux connector for use with any component compatible with TProps.
  This connector also accepts props of type TOwnProps.
  Will return an empty react node at runtime if used without children, with an
  error message if compiled with -debug.
*/
typedef ReactGenericConnectorOfProps<TProps, TOwnProps> = ReactConnect<TVoid, TProps, TOwnProps>;
```

### General notes

 * Explicit typing is needed for `mapStateToProps` and `mapDispatchToProps` functions, to allow macros to check typing
 * There is no way of knowing you application store's state type, so appState is typed as `Dynamic`
 * Connectors cannot use a functional component as default component, unless it's a `@:jsxStatic` class
 * Connectors handle `defaultProps` when they have ownProps
 * `displayName`s are handled, to allow easier debugging with react extension

 * React-redux defaults to `(dispatch) => ({dispatch})` for `mapDispatchToProps` when you pass `null`, meaning `dispatch` will be provided if you do not define a `mapDispatchToProps` returning an object. This behavior has *not* been reproduced here, instead the macro will use a function returning an empty object.

 * This PR needs some PRs from haxe-react:
	 * [#86](https://github.com/massiveinteractive/haxe-react/pull/86) for its JsxStaticMacro refactoring
	 * [#96](https://github.com/massiveinteractive/haxe-react/pull/96) for `ReactUtil.copyWithout()`
	 * [#97](https://github.com/massiveinteractive/haxe-react/pull/97) for pluggable ReactComponent build macros (for `@:connect`)

### Example app and Tests

The example app has been updated for this PR: https://github.com/kLabz/haxe-redux-todo/tree/redux/6

Tests have been added, using `enzyme`. You can find the tests [here](https://github.com/kLabz/haxe-redux-todo/blob/redux/6/src/test/suite/ReactConnectorTests.hx).



## `@:connect` meta on ReactComponent

Add support for `@:connect` meta on react components, as the `@connect` decorator of react-redux.
Note that typing is not checked here; however react-redux does some checks at runtime.

### `@:connect` without params

If you do not provide params to the connect meta, a macro will loop through your fields to find *static* fields named `mapStateToProps`, `mapDispatchToProps`, `mergeProps` or `options` and use those it found.

```haxe
@:connect
class MyComponent extends ReactComponentOfProps<Props>
{
	static function mapDispatchToProps(dispatch:Dispatch):Props
	{
		return {
			onClick: function() return dispatch(Actions.Click)
		};
	}

	override public function render()
	{
		return jsx('
			<button onClick=${props.onClick} />
		');
	}
}
```

### Providing params

You can use any function instead of your own static fields.

```haxe
@:connect(null, Helpers.myMapDispatchToProps)
class MyComponent extends ReactComponentOfProps<Props> {}
```

Or even declare your function inline, if you're into this:

```haxe
@:connect(null, function(dispatch:Dispatch) return {onClick: function() return dispatch(Actions.Click))
class MyComponent extends ReactComponentOfProps<Props> {}
```

Please note that you must respect the order of arguments of `ReactRedux.connect()`, so you must pass null as first argument if you only want to define `mapDispatchToProps`, for example.

kLabz committed:
Implemented ReactConnector (from ReactRedux.connect())

Documentation: added example
Improved @:connect type detection and errors
Use @:jsxStatic for containers
Allow using both dispatch and ownProps in @:connect functions

kLabz commented:
## Getting rid of `<MyContainer.Connected />`

A solution to be able to use `MyContainer` in JSX could be a patch to [haxe-react](https://github.com/massiveinteractive/haxe-react)'s parseJsx macro:

* In `ReactMacro.parseJsxNode`, case `case JsxAst.Node`:
 Add `parseJsxType(type);` after `type.pos = pos;`

* Add `ReactMacro.parseJsxType`:
```haxe
static function parseJsxType(type:Expr)
{
	var typedExpr = Context.typeExpr(type);

	switch (typedExpr.expr)
	{
		case TTypeExpr(TClassDecl(_.get() => c)):
			var superClass = c.superClass;
			if (superClass != null)
			{
				var superClassType = c.superClass.t.get();
				if (superClassType.name == 'ReactContainerFor')
				{
					var pack = superClassType.pack;
					if (pack.length == 2 && pack[0] == 'redux' && pack[1] == 'react')
					{
						type.expr = EField({
							expr: EConst(CIdent(c.name)),
							pos: type.pos
						}, 'Connected');
					}
				}
			}

		default:
	}
}
```

This would add the `.Connected` at compile-time.

elsassph commented on redux/react/ConnectedComponentMacro.hx:
```diff
@@ -0,0 +1,665 @@
+		else if (connectedFunctions.length > 0)
+			addMapDispatchToPropsWarnings(connectedFunctions);
+
+		if (mapStateToProps == null) addMapStateToProps(fields, inClass);
```
We should avoid creating all the extra functions.

kLabz replied:
addMapStateToProps, addMergeProps and addOptions only create `null` objects so that 
```
redux.react.ReactRedux.connect(mapStateToProps, mapDispatchToProps, mergeProps, options)(render);
```
does not get any `undefined` (you can see them at the very bottom of the file)

elsassph commented on redux/react/ConnectedComponentMacro.hx:
```diff
@@ -0,0 +1,665 @@
+
+	static function renderExpr(componentName:String, pos:Position)
+	{
+		var jsxStr = '<$' + componentName + ' {...props} />';
```
`$` is not needed  - that's really only a convenience for "click to navigate" in editors.
I had some doubts but I think you can pass children this way as well.

elsassph commented:
I've also started to accept that the separated connected HOC is ultimately better...

I like the trick to avoid `.Connected`.

`ReactContainerFor<T>` could be renamed `ReactConnect<T>`.

Possibly `@:connect` isn't even needed: static functions with `dispatch` as 1st argument could be automatically connected.

But very bad example: `VisibleTodoList` should NOT filter the todos, UNLESS you can guarantee that for the same combination of todos and filter you'll get the same array references (`.filter` creates new arrays each time), otherwise any unrelated state change change will trigger re-renderers of the `TodoList`. Typically, filtering should happen in the render function.

Yes we want the `Partial` to be in `haxe-react`.

kLabz commented:
Thanks for your feedback :)

I removed the unneeded $ and renamed `ReactContainerFor<T>` to `ReactConnect<T>`.

> Possibly @:connect isn't even needed: static functions with dispatch as 1st argument could be automatically connected.

I agree that this would be easier to use.

However, it could generate more runtime (and possibly obscure) errors since some functions would be discarded by the macro when the user might think it is being used (if first argument counts on type inference to be of type "Dispatch", the macro won't know that the function needs to be "connected").

I'm not sure what to do here, both solution have pros and cons.
Do you want to check the first argument instead of the meta?

> But very bad example: VisibleTodoList should NOT filter the todos

I agree, but it seems weird that the official example from redux documentation would have such a flaw. I'll look into it, maybe this is one of the "many useful optimizations" done behind the whole `connect()` process.

kLabz commented:
Not ready to be merged. Will be updated with an example project in a few days.

kLabz commented:
Created an [example app](https://github.com/kLabz/haxe-redux-todo), ported from react-redux doc. I added the original javascript sources as comments to emphasize the similarities between the two syntaxes.

Known issues for this PR, that still needs some more work:
 * No support for automatic `mapDispatchToProps` (via @:connect) with `dispatch` and `ownProps` as arguments. At the moment, we need to implement `mapDispatchToProps` ourselves to achieve this ([example](https://github.com/kLabz/haxe-redux-todo/blob/master/src/containers/FilterLink.hx)).
 This feature will make the automatic matching based on the arguments more difficult to implement (but updating @:connect shouldn't be too hard).
 * Some errors needs more helpful messages. I'll try to list them and handle them asap.

Overall, it's usable but edge cases may cause some issues which could be hard to handle because of the lack of proper error messages.

elsassph commented:
Nice sample.

I think we need to generalize the system to get rid of `.Connected` of `<MyContainer.Connected />`. React lib shouldn't know about redux. Maybe a meta would be enough. 

Why generalize: I'd like a cleaner way to make functional components.
```haxe
@:jsxStatic('render') // could be added by a ReactFunction interface for instance
class Link {
    static function render(props:LinkProps) {
        return jsx('<a href={props.to}>{props.children}</a>
    } 
}

// usage
function render() {
     return jsx('<Link to="abc">Abc</Link>');
}

// replaced by
function render() {
     return jsx('<Link.render to="abc">Abc</Link.render>');
}
```

kLabz commented:
I added a PR to haxe-react implementing `@:jsxStatic` and updated this one.

I'll look into the other known issues and hope to finish this PR soon, at last :)

elsassph commented:
Can you update the PR description accordingly?

Also still thinking about the `@:connect` in the `ReactContainerFor`, what about:
```haxe
@:dispatch
function onTodoClick(id:Int)
{
	return TodoAction.Toggle(id);
}
```
Which would be wrapped in a little function that would call dispatch (we'd want to be careful not creating a new function on each update).

kLabz commented:
I was not happy with what I had, so I partially rewrote this PR to (IMO anyway) improve its core functionalities.

I updated the PR description and the example app, the main differences being using another name and allowing to easily reuse a HOC with another (or even multiple) wrapped component.



As for `@:dispatch`, I'm really not sure. 

It looks a lot like using `mapDispatchToProps()`, which is available in this PR too (it's actually the default behavior, `@:connect` just being a shortend). Also, sometimes we need both `dispatch` and `ownProps`, which would require another meta.

About creating a new function on each update, it seems to be happening in both this implementation (either with mapDispatchToProps or with @:connect) and the javascript redux examples. But `mapDispatchToProps` is only called when the HOC's props change, so it's really a tiny issue.

kLabz committed:
ReactConnector: handle displayName if -debug

zabojad commented:
Guys, what's the status with this PR? Are we close to an agreement :)?

elsassph commented:
Can we avoid the `static var wrappedComponent:Link`? 
For instance as `class FilterLink extends ReactConnectorOf<Link, LinkProps, FilterLinkProps>`

kLabz commented:
Yeah, I don't like it either. I have another improvement to push, too.

kLabz committed:
ReactConnector fixes and improvements

kLabz commented:
I updated the PR (and its description), with many fixes / improvements and some tests (available [in the example app](https://github.com/kLabz/haxe-redux-todo/blob/redux/6/src/test/suite/ReactConnectorTests.hx)). This feels a lot more like a finished feature now :)

Main changes: 
 * ditched the whole @:connect / @:dispatch thing, since it doesn't bring much except problems :D
 * replaced `wrappedComponent` by `TComponent` param, as suggested by @elsassph 
 * added / consolidated HOC composition, defaultProps support, displayName, etc.

Note that this PR needs two PRs from haxe-react:
 * [#86](https://github.com/massiveinteractive/haxe-react/pull/86) for its JsxStaticMacro refactoring
 * [#96](https://github.com/massiveinteractive/haxe-react/pull/96) for `ReactUtil.copyWithout()`

kLabz commented:
Digging deeper, I'll have to make an update tomorrow to support factories, too:
> If the mapStateToProps argument supplied to connect returns a function instead of an object, it will be used to create an individual mapStateToProps function for each instance of the container.

This is especially useful when using reselect (more on using reselect with haxe later).

_Damn javascript libs with their infinite and not often documented signatures x_x_

kLabz committed:
Updated ReactRedux.connect() signature

kLabz committed:
Support for @:connect on ReactComponent

kLabz commented:
I updated `ReactRedux.connect()` signature to handle factories and action creators.

While checking react-redux tests, I noticed the `@connect` decorator available when using redux in javascript, so I added a similar feature for haxe:

## `@:connect` meta on ReactComponent

Add support for `@:connect` meta on react components, as the `@connect` decorator of react-redux.
Note that typing is not checked here; however react-redux does some checks at runtime.

### `@:connect` without params

If you do not provide params to the connect meta, a macro will loop through your fields to find *static* fields named `mapStateToProps`, `mapDispatchToProps`, `mergeProps` or `options` and use those it found.

```haxe
@:connect
class MyComponent extends ReactComponentOfProps<Props>
{
	static function mapDispatchToProps(dispatch:Dispatch):Props
	{
		return {
			onClick: function() return dispatch(Actions.Click)
		};
	}

	override public function render()
	{
		return jsx('
			<button onClick=${props.onClick} />
		');
	}
}
```

### Providing params

You can use any function instead of your own static fields.

```haxe
@:connect(null, Helpers.myMapDispatchToProps)
class MyComponent extends ReactComponentOfProps<Props> {}
```

Or even declare your function inline, if you're into this:

```haxe
@:connect(null, function(dispatch:Dispatch) return {onClick: function() return dispatch(Actions.Click))
class MyComponent extends ReactComponentOfProps<Props> {}
```

Please note that you must respect the order of arguments of `ReactRedux.connect()`, so you must pass null as first argument if you only want to define `mapDispatchToProps`, for example.

elsassph commented:
Re "Reusing the HOC with another wrapped component": not sure to understand, a connected component can have children like any component - it shouldn't replace the wrapped component.

Like `<FilterLink filter="SHOW_ALL">Show all</FilterLink>`

kLabz commented:
I guess this behavior should only happen with `ReactGenericConnector` and `ReactGenericConnectorOfProps`, since they are the only ones really needing it (and the only cases where it actually makes sense).

There were initially no "generic" connectors, so I applied the rule to all connectors but that shouldn't be the case any more.

I should modify the behavior so that:
 * "normal" containers pass `props.children` to their wrapped component (`TComponent`)
 * "normal" containers are deeply tied with their `TComponent` and will only work (provide the generated props) with this component and no other component
 * only `ReactGenericConnector` and `ReactGenericConnectorOfProps` can be used with any component, and will provide the generated props to their children. They are more providers than classical HOCs, except that they provide props and not context.

_Note: `TComponent` is the wrapped component, as in `connect(...)(TComponent)`_

kLabz commented:
Extracted `@:connect` in #14, this should handle most of the cases (but does not provide type checks) so that this PR can wait.
